### ISSUE TITLE
Rule Request: Inert defer

### ISSUE THREAD

**Issue by jpsim** [realm/SwiftLint#2123]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Rule Request

> 1. Why should this rule be added? Share links to existing discussion about what
>    the community thinks about this.

I recently introduced a bug in some code that used `differ`.

**Before:**

```swift
func foo() {
  defer { /* deferred code */ }
  // more code
}
```

**After:**

```swift
func foo() {
  if condition {
    defer { /* deferred code */ }
  }

  // more code
}
```

My intention was to only run the deferred code if the condition was true. What happened in practice is that the deferred code was run at the end of the scope of the if statement body. 🙀 

If we had a rule that warned if there was an inert `defer` statement, I would have realized my sin immediately, rather than ship a bug only to be caught later.

> 2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

```swift
// Would trigger

func example1() {
  defer { /* deferred code */ }
  // comment
}

func example2() {
  if condition {
    defer { /* deferred code */ }
    // comment
  }
}

// Wouldn't trigger

func example3() {
  defer { /* deferred code */ }
  
  print("other code")
}

func example4() {
  if condition {
    defer { /* deferred code */ }
    print("other code")
  }
}

```

> 3. Should the rule be configurable, if so what parameters should be configurable?

I don't think this needs to be configurable.

> 4. Should the rule be opt-in or enabled by default? Why?
>    See [README.md](../README.md#opt-in-rules) for guidelines on when to mark a
>    rule as opt-in.

Enabled by default.

**Comment by SDGGiesbrecht** [comment]
So, in essence:

“If defer is at the end of its parent scope, it will be executed right where it is anyway, so throw a warning.”

I like it.

**Comment by SDGGiesbrecht** [comment]
P.S. I’ve been bitten by that same mistake myself more than once...

**Comment by harlanhaskins** [comment]
IMO this is compiler warning territory.

**Comment by keith** [comment]
Joe was nice enough to make a swift ticket for this https://bugs.swift.org/browse/SR-7307

**Comment by jpsim** [comment]
Agreed this should live in the compiler. Doesn't mean it can't also live in SwiftLint, so folks who can't upgrade to the future Swift version that would include that diagnostic could still get it via SwiftLint.

**Comment by marcelofabri** [comment]
https://bugs.swift.org/browse/SR-7307 was implemented 🎉

**Comment by jpsim** [comment]
Yes it should be possible to detect. Could you please file an issue tracking that work?

**Comment by rzulkoski** [comment]
> Yes it should be possible to detect. Could you please file an issue tracking that work?

@darecki @jpsim  Did this issue ever get created?

**Comment by kubatruhlar** [comment]
@darecki  Not only initializers but also `deinit` has the same behaviour. Check [SR-10207](https://bugs.swift.org/browse/SR-10207) for more.
